explorer/events.go

```diff
@@ -131,144 +95,234 @@ func AppliedEvents(cs consensus.State, b types.Block, cu ChainUpdate) []Event {
-					FileContract:  fce,
-					SiacoinOutput: sces[outputID],
-					Missed:        false,
-				}, []types.Address{fce.FileContract.ValidProofOutputs[i].Address})
+				address := fce.FileContract.ValidProofOutputs[i].Address
+				element := sces[types.FileContractID(fce.ID).ValidOutputID(i)]
+
+				addEvent(types.Hash256(element.ID), element.MaturityHeight, wallet.EventTypeV1ContractResolution, EventV1ContractResolution{
+					Parent:         efc,
+					SiacoinElement: SiacoinOutput{SiacoinElement: element},
+					Missed:         false,
+				}, []types.Address{address})
 			}
 		} else {
 			for i := range fce.FileContract.MissedProofOutputs {
-				outputID := fce.ID.MissedOutputID(i)
-				addEvent(types.Hash256(outputID), cs.MaturityHeight(), &EventContractPayout{
-					FileContract:  fce,
-					SiacoinOutput: sces[outputID],
-					Missed:        true,
-				}, []types.Address{fce.FileContract.MissedProofOutputs[i].Address})
+				address := fce.FileContract.MissedProofOutputs[i].Address
+				element := sces[types.FileContractID(fce.ID).MissedOutputID(i)]
+
+				addEvent(types.Hash256(element.ID), element.MaturityHeight, wallet.EventTypeV1ContractResolution, EventV1ContractResolution{
+					Parent:         efc,
+					SiacoinElement: SiacoinOutput{SiacoinElement: element},
+					Missed:         true,
+				}, []types.Address{address})
 			}
 		}
 	})
 
+	cu.ForEachV2FileContractElement(func(fce types.V2FileContractElement, _ bool, rev *types.V2FileContractElement, res types.V2FileContractResolutionType) {
+		if res == nil {
+			return
+		}
+
+		fce.StateElement.MerkleProof = nil
+
+		var missed bool
+		if _, ok := res.(*types.V2FileContractExpiration); ok {
+			missed = true
+		}
+
+		var typ string
+		switch res.(type) {
+		case *types.V2FileContractRenewal:
+			typ = "renewal"
+		case *types.V2StorageProof:
+			typ = "storageProof"
+		case *types.V2FileContractExpiration:
+			typ = "expiration"
+		default:
+			panic("unknown resolution type")
+		}
+
+		efc := V2FileContract{V2FileContractElement: fce}
+		{
+			element := sces[types.FileContractID(fce.ID).V2HostOutputID()]
+			addEvent(types.Hash256(element.ID), element.MaturityHeight, wallet.EventTypeV2ContractResolution, EventV2ContractResolution{
+				Resolution: V2FileContractResolution{
+					Parent:     efc,
+					Type:       typ,
+					Resolution: res,
+				},
+				SiacoinElement: SiacoinOutput{SiacoinElement: element},
+				Missed:         missed,
+			}, []types.Address{fce.V2FileContract.HostOutput.Address})
+		}
+
+		{
+			element := sces[types.FileContractID(fce.ID).V2RenterOutputID()]
+			addEvent(types.Hash256(element.ID), element.MaturityHeight, wallet.EventTypeV2ContractResolution, EventV2ContractResolution{
+				Resolution: V2FileContractResolution{
+					Parent:     efc,
+					Type:       typ,
+					Resolution: res,
+				},
+				SiacoinElement: SiacoinOutput{SiacoinElement: element},
+				Missed:         missed,
+			}, []types.Address{fce.V2FileContract.RenterOutput.Address})
+		}
```

Comment:
We usually don't use scopes like this so I'd avoid them for amore consistent code style across our projects.
e.g. in this case an alternative that provides similar safety with less copy-paste risk would be 

```go
		addV2Resolution := func(element types.SiacoinElement) {
			efc := V2FileContract{V2FileContractElement: fce}
			addEvent(types.Hash256(element.ID), element.MaturityHeight, wallet.EventTypeV2ContractResolution, EventV2ContractResolution{
				Resolution: V2FileContractResolution{
					Parent:     efc,
					Type:       typ,
					Resolution: res,
				},
				SiacoinElement: SiacoinOutput{SiacoinElement: element},
				Missed:         missed,
			}, []types.Address{fce.V2FileContract.HostOutput.Address})
		}
		addV2Resolution(sces[types.FileContractID(fce.ID).V2RenterOutputID()])
		addV2Resolution(sces[types.FileContractID(fce.ID).V2HostOutputID()])
```